backend/app/api/routes/dde.py

```diff
@@ -214,33 +216,121 @@ async def extract_from_file(*, session: SessionDep, current_user: CurrentUser, n
+                for value in extracted_data if isinstance(value, (int, float)))
+
+def find_key_by_value(combined_token, extracted_data) -> str | None: 
+    try:
+        combined_token = float(combined_token)
+    except ValueError:
+        return None
+    return next((k for k, v in extracted_data.items() 
+                    if isinstance(v, (int, float)) and is_equal_ignore_sign(combined_token, v)), None)    
+
+def extract_logprobs_from_response(response: Any, extracted_data: dict[str, Any]) -> dict[str, float]:
+    logprob_data = {}
+    tokens_info = response.choices[0].logprobs.content
+
+    def process_numeric_values(extracted_data: dict, path=''):
+        i = 0
+        while i < len(tokens_info):    
+            token = tokens_info[i].token
+            
+            if token.isdigit():          # Only process tokens that are numeric
+                combined_token, combined_logprob = combine_tokens(tokens_info, i)
+                
+                if combined_token_in_extracted_data(combined_token, extracted_data.values()):     #Checks if a combined token matches any numeric values in the extracted data.
+                    key = find_key_by_value(combined_token, extracted_data)                  #Finds the key in 'extracted_data' corresponding to a numeric value that matches the combined token.
+                    if key:
+                        full_key = path + key  
+                        logprob_data[full_key + '_prob_first_token'] = math.exp(tokens_info[i].logprob)
+                        logprob_data[full_key + '_prob_second_token'] = math.exp(tokens_info[i+1].logprob)
+
+                        toplogprobs_firsttoken = tokens_info[i].top_logprobs
+                        toplogprobs_secondtoken = tokens_info[i+1].top_logprobs
+
+                        logprobs_first = [top_logprob.logprob for top_logprob in toplogprobs_firsttoken]
+                        logprobs_second = [top_logprob.logprob for top_logprob in toplogprobs_secondtoken]
+
+                        logprob_data[full_key + '_first_token_toplogprobs'] = logprobs_first
+                        logprob_data[full_key + '_second_token_toplogprobs'] = logprobs_second
+            i += 1
+
+    def traverse_and_extract(data, path=''):
+        for key, value in data.items():
+            if isinstance(value, dict):
+                traverse_and_extract(value, path + key + '.')
+            elif isinstance(value, (int, float)):
+                process_numeric_values(data, path)
+    
+    traverse_and_extract(extracted_data)
+    
+    return logprob_data
+
+
+def combine_tokens(tokens_info: list[dict[str, Any]], start_index: int) -> tuple[str, float]: 
+    combined_token = tokens_info[start_index].token
+    combined_logprob = tokens_info[start_index].logprob
 
+    i = start_index
+    # Keep combining tokens as long as the next token is a digit
+    while i + 1 < len(tokens_info) and tokens_info[i + 1].token.isdigit():
```

Comment:
Why not use a "for loops" ? "While loops" may never end and it is not always straightforward to check.